src/macros-by-example.adoc

```diff
@@ -0,0 +1,823 @@
+container with a macro invocation.  The text form mirrors that limitation, but you can always
+call a macro inside the delimiter, with no loss of generality:
+
+----
+(:prices [(:values 10)] GBP) ⇒ {amount:10, currency:GBP}
+----
+
+Note that the delimiter is required even for single values to avoid ambiguity.  Consider this macro:
+
+[{mrk}]
+----
+(*public* ouch [(*list** stuff)] …)
+----
+
+In the E-expression `(:ouch [])`  is that list the delimiting container, or the sole input value?
+When expanding the E-expression does `stuff` produce void, or a single empty list?  `***`-arguments
+require a delimiting container precisely so there’s no ambiguity, regardless of the element type,
+and the answer is that `stuff` is void.
+
+Within the delimiting container, the invocation can have any number of values and/or macro
+invocations.  The macro parameter produces the results of those expressions, concatenated into a
+single stream, and the expander verifies that each value on that stream is acceptable by the
+parameter’s declared type.
+
+----
+(:prices (1 (:values 2 3) 4) GBP) ⇒ {amount:1, currency:GBP}
+                                    {amount:2, currency:GBP}
+                                    {amount:3, currency:GBP}
+                                    {amount:4, currency:GBP}
+----
+
+
+==== one-or-more
+
+A one-or-more `*+*` parameter works just like `***` except the resulting stream must produce at
+least one value.  To continue using our prices example:
+
+[{mrk}]
+----
+(*public* prices
+  [(*number+* amount), (*symbol* currency)]
+  (*for_each* amount
+    (price amount currency)))
+----
+
+[{mrk}]
+----
+(:prices () JPY) ⇒ _**error**: at least one value expected for + parameter_
+(:prices [99] EUR)               ⇒ {amount:99, currency:EUR}
+(:prices \((:) (:values 99)) EUR) ⇒ {amount:99, currency:EUR}
+----
+
+Note that among the delimited expressions, empty results are fine, as long as their concatenation
+produces one or more values.
+
+
+=== Optional Arguments
+
+When a trailing parameter is voidable, an invocation can omit its corresponding argument, as long
+as no following parameter is being given a value.  We’ve seen this as applied to rest-parameters,
+but it also applies to `*?*` and `***` cardinality:
+
+[{mrk}]
+----
+(*public* optionals
+  [(**any* **a), (*any?* b), (*any!* c), (*any** d), (*any?* e), (*any\...* f)]
+  (make_list a b c d e f))
+----
+
+Since `d`, `e`, and `f` are all voidable, they can be omitted by invokers.  But `c` is required so
+`a` and `b` must always be present, at least as an empty delimiter:
+
+----
+(:optionals [] (:) for_c) ⇒ [for_c]
+----
+
+Now `c` receives the symbol `for_c` while the other parameters are all void.  If we want to provide
+just `e`, then we must also provide a delimiter for `d`:
+
+----
+(:optionals [] (:) for_c () for_e) ⇒ [for_c, for_e]
+----
+
+
+=== Tagless and Fixed-Width Types
```

Comment:
Did we consider the encoder inferring that a type _may_ be tagless and just, well, not tagging it?

I suppose we'd need a way to restrict that the type was not just a non-null "concrete" type _and_ that it didnt' allow annotations... but I'd say if something can be made tagless we should just do that for folks. Adding a modifier for them to specify it just adds the complexity of them taking advantage of it when they can.

Reply:
IIUC the binary encoding has a way to "remove tags on the fly" when it makes sense to do so.  I'll let @zslayton explain, but I think the summary is that when there's more than one value provided to a tagged parameter, and all values can be encoded without tags, then a bit is set and the tags are omitted for that invocation.

The `tagless` types basically require the data to follow that constraint, so that additional overhead can be avoided.

Reply:
In talking to Zack, the other restriction (that I think is called out in a different section) is that an argument passed as a tagless parameter cannot be a macro invocation.